### ISSUE TITLE
Proposal: Make Extensions.Mapping.FunctionMapping.index optional

### ISSUE THREAD

**Issue by westonpace** [substrait-io/substrait#69]
This is continuing the discussion from #65.

Rationale:

* Given the input types and the output type then the variant for a given function expression could be automatically detected by a validator (or the call could be found to be non conformant because no variant exists which satisfies the criteria).
* It's hard to compute (see #65)
* Function variants should not change the semantic meaning of the function.  Therefore, if producer and consumer agree on input types, output type, and (extension namespaced) function name then they agree on a common semantic meaning.  Adding the variant index does not help in this process.

> Lastly, I would add that I think most systems will either: declare static mappings between internal operations and semantic definitions or do build-time dynamic binding (as opposed to doing so at runtime). The changes here also make that more straightforward, explicit and reliable.
@jacques-n (quote pulled from discussion on #65)

This is exactly what I am doing.  Here is my current implementation:

```
            var func = new Expression.Types.ScalarFunction
            {
                // FunctionIdFromName is (more or less) a static mapping from function name to a function id
                Id = new Extensions.Types.FunctionId {Id = SubstraitUtil.FunctionIdFromName(functionName)},
                OutputType = SubstraitUtil.SubstraitTypeFromCSharpType(expr.Type)
            };
            func.Args.Add(leftArg);
            func.Args.Add(rightArg);
```

If I have to add a variant index using the old yaml definition the inner function creation expands with hard-coded variant ids:

```
            var func = new Expression.Types.ScalarFunction
            {
                OutputType = SubstraitUtil.SubstraitTypeFromCSharpType(expr.Type)
            };
            if (expr.Type.IsAssignableTo(typeof(decimal)))
            {
                func.Id = SubstraitUtil.FunctionIdFromName(functionName, /*variant_idx=*/1);
            }
            else
            {
                func.Id = SubstraitUtil.FunctionIdFromName(functionName, /*variant_idx=*/0);
            }
            func.Args.Add(leftArg);
            func.Args.Add(rightArg);
```

Under the new YAML proposed it gets even more onerous:

```
            var func = new Expression.Types.ScalarFunction
            {
                OutputType = SubstraitUtil.SubstraitTypeFromCSharpType(expr.Type)
            };
            int variantIndex = 0;
            switch (func.OutputType.KindCase)
            {
                case Type.KindOneofCase.I8:
                    variantIndex = 1;
                    break;
                case Type.KindOneofCase.I16:
                    variantIndex = 2;
                    break;
                case Type.KindOneofCase.I32:
                    variantIndex = 3;
                    break;
                case Type.KindOneofCase.I64:
                    variantIndex = 4;
                    break;
                case Type.KindOneofCase.Fp32:
                    variantIndex = 5;
                    break;
                case Type.KindOneofCase.Fp64:
                    variantIndex = 6;
                    break;
                case Type.KindOneofCase.Decimal:
                    variantIndex = 7;
                    break;
                default:
                    throw new Exception(
                        $"Illegal arithmetic variant: functionName={functionName} outputType={expr.Type}");
            }
            func.Id = SubstraitUtil.FunctionIdFromName(functionName, variantIndex);
            func.Args.Add(leftArg);
            func.Args.Add(rightArg);
```

And this is assuming the variant ordering is consistent between all the different arithmetic functions.  Indeed, at this point, this seems brittle enough and error-prone enough that I would probably go through the bother of doing the output type derivation myself so I could compute the variant index that way.

In summary, I'm doing a lot more work, but I don't see what I'm gaining.

**Comment by jacques-n** [comment]
I think you misinterpreted the proposal in #65. It proposes the removal of the index and a replacement of a naming scheme. For example `add:opt_i8_i8` as described [here](https://gist.github.com/jacques-n/86e4675aa076c65b8004572e6952f1bc#file-scalar_functions-yaml-L8). The goal as part of #65 (and in general) is that plans have self evident variant specific bindings (and consumers do not have to have rules around how to pick which variant). This keeps consumers simpler and encapsulates all "selection logic" in a producer. 

We considered the type names to be implicit in the argument types but ultimately felt like that was over-simplification and also disallowed the possibility of overloads. For example, having a plan clearly state the use of `add(i8,i8)` versus `add(any, any)`.

**Comment by cpcloud** [comment]
So, playing devil's advocate here, one way to address the `any` problem is to add an `Any` type to the type system. Since expressions have type information in them, baking that into the function name is redundant, perhaps not onerously so but as opposed to handling variants it seems much more desirable to handle a _single_ name to a _single_ id.

**Comment by cpcloud** [comment]
Personally, I think if there were away to avoid forcing all producers and consumers to enumerate function_ids in the same way, that is, `0` in the producer is the same as `0` in the consumer, that would be ideal.

**Comment by jacques-n** [comment]
> So, playing devil's advocate here, one way to address the any problem is to add an Any type to the type system.

I'm not sure how this fixes the disambiguation problem. Unless you're saying that in order to use a Any function, you must cast the arguments to any as well, which seems pretty confusing.

> I think if there were a way to avoid forcing all producers and consumers to enumerate function_ids in the same order, that is, 0 in the producer is the same as 0 in the consumer, that would be ideal.

I don't feel like that has ever been forced (if I follow what you're saying). I'd expect that a producer could send query 1 with add:i8_i8 bound to functionId:0 and then later send it as functionId:4. On the receiving side, you decode using the function name in the function mapping, the id is just to normalize the information to the name.

**Comment by cpcloud** [comment]
> Unless you're saying that in order to use a Any function, you must cast the arguments to any as well, which seems pretty confusing.

Yeah, I don't think I thought very far along here. That would be confusing.

>  the id is just to normalize the information to the name.

Ah, I didn't realize that it was ephemeral (or, tied to the lifetime of a plan). In that case, I am fine with @jacques-n's proposal.

@westonpace Does the id being ephemeral address your point about brittleness?

**Comment by jacques-n** [comment]
> Does the id being ephemeral address your point about brittleness?

I think @westonpace was referring to [this field](https://github.com/substrait-io/substrait/blob/main/binary/extensions.proto#L50), which #65 also is trying to remove.

**Comment by westonpace** [comment]
> @westonpace Does the id being ephemeral address your point about brittleness?
> I think @westonpace was referring to this field, which #65 also is trying to remove.

@jacques-n is correct.  Generating a string like `_ii` which is just a re-encoding of the input (and output?) arguments is not too onerous (although I may still have special cases for the wildcard options but hopefully there will be few of those).

It was not clear to me that #65 was dropping the index field but that solve this.  I'll close this and go back to commenting on #65 .
